### PR TITLE
Added binary size test

### PR DESCRIPTION
**Pull request by andrei-cosmin**
## Reason for This PR

Issue #1337

## Description of Changes

Added a test to check if the size of the release binary is within expected range

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria. Where there are two options, keep one.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] Either this PR is linked to an issue, or, the reason for this PR is
      clearly provided.
- [x] The description of changes is clear and encompassing.
- [x] Either no docs need to be updated as part of this PR, or, the required
      doc changes are included in this PR. Docs in scope are all `*.md` files
      located either in the repository root, or in the `docs/` directory.
- [x] Either no code has been touched, or, code-level documentation for touched
      code is included in this PR.
- [x] Either no API changes are included in this PR, or, the API changes are
      reflected in `firecracker/swagger.yaml`.
- [x] Either the changes in this PR have no user impact, or, the changes in
      this PR have user impact and have been added to the `CHANGELOG.md` file.
- [x] Either no new `unsafe` code has been added, or, the newly added `unsafe`
      code is unavoidable and properly documented.

### CODE REVIEW

File: tests/integration_tests/build/test_binary_szie.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
+# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+"""Test that checks if the release binary size is within expected size"""
+
+import os
+import pytest
+
+import host_tools.cargo_build as host
+
+BINARY_SIZE_TARGET = 3137280
```
Comment (acatangiu) [inline tests/integration_tests/build/test_binary_szie.py]: The `firecracker` binary is where we're really interested, but please also add a check for the `jailer` binary size.

Comment (raduiliescu) [reply tests/integration_tests/build/test_binary_szie.py]: Do you have the same binary size for both ARM and X86?

File: tests/integration_tests/build/test_binary_szie.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    "features",
+    FEATURES
+)
+@pytest.mark.timeout(500)
+def test_binary_size(test_session_root_path, features):
+    """
+    This will check if the size of the release binary is within expected range
+    """
+
+    fc_binary, _ = host.get_firecracker_binaries(
```
Comment (aghecenco) [inline tests/integration_tests/build/test_binary_szie.py]: Does this rebuild the binaries when the first microvm is instantiated?

File: tests/integration_tests/build/test_binary_szie.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
+# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+"""Test that checks if the release binary size is within expected size"""
```
Comment (aghecenco) [inline tests/integration_tests/build/test_binary_szie.py]:
```suggestion
"""Tests that check if the release binary sizes fall within expected ranges."""
```

File: tests/integration_tests/build/test_binary_szie.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+FEATURES = [""]
+
+@pytest.mark.parametrize(
+    "features",
+    FEATURES
+)
+@pytest.mark.timeout(500)
+def test_binary_size(test_session_root_path, features):
+    """
+    This will check if the size of the release binary is within expected range
```
Comment (aghecenco) [inline tests/integration_tests/build/test_binary_szie.py]:
```suggestion
    Test if the size of the release binaries are within expected ranges.
```

File: tests/integration_tests/build/test_binary_szie.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    binary_low_msg = (
+        'Current binary size of {} bytes is below the target of {} bytes with {} bytes.\n'
+        'Perhaps the value of {} should be updated'
+        .format(binary_size, BINARY_SIZE_TARGET, BINARY_SIZE_TARGET - binary_size, binary_size_target_name)
+    )
+
+    assert binary_size > BINARY_SIZE_TARGET - BINARY_SIZE_DELTA, binary_low_msg
+
+    binary_high_msg = (
+        'Current binary size {} is above the target of {} bytes with {} bytes.\n'
```
Comment (aghecenco) [inline tests/integration_tests/build/test_binary_szie.py]: Please keep the wording consistent with `binary_low_msg`.

**Comment by sandreim** [comment]
Is there any value in keeping size history across versions or PRs?

File: tests/integration_tests/build/test_binary_szie.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+import os
+import pytest
+
+import host_tools.cargo_build as host
+
+BINARY_SIZE_TARGET = 3137280
+"""Target binary size to be checked in bytes"""
+
+BINARY_SIZE_DELTA = 1024 * 1024 / 2
```
Comment (sandreim) [inline tests/integration_tests/build/test_binary_szie.py]: I would use a percentage here. Also 0.5MB is a lot, we do not usually have PRs that are that beefy in terms of binary output, unless we add extra dependencies which we should also avoid.

File: tests/integration_tests/build/test_binary_size.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+import pytest
+import platform
+
+import host_tools.cargo_build as host
+
+MACHINE = platform.machine()
+""" Platform definition used to select the correct size target"""
+
+FC_BINARY_SIZE_TARGET = 3137280 if MACHINE == "x86_64" else 3536544
+"""Firecracker target binary size to be checked in bytes"""
```
Comment (aghecenco) [inline tests/integration_tests/build/test_binary_size.py]: nit
```suggestion
"""Firecracker target binary size in bytes"""
```

File: tests/integration_tests/build/test_binary_size.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+JAILER_BINARY_SIZE_TARGET = 2985152 if MACHINE == "x86_64" else 3149128
+"""Jailer target binary size to be checked in bytes"""
```
Comment (aghecenco) [inline tests/integration_tests/build/test_binary_size.py]: nit
```suggestion
"""Jailer target binary size in bytes"""
```

File: tests/integration_tests/build/test_binary_size.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+BINARY_SIZE_TOLERANCE = 0.05
```
Comment (aghecenco) [inline tests/integration_tests/build/test_binary_size.py]: Should we have a hard limit? If we gradually fatten up the binary, we might not notice if it gets excessively big.

File: tests/integration_tests/build/test_binary_size.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+"""Tolerance of 5% allowed for binary size"""
+
+FEATURES = [""]
```
Comment (aghecenco) [inline tests/integration_tests/build/test_binary_size.py]: We don't have any features, might as well remove this and the parametrization to keep the code as simple as possible. We can always re-add it if we feature-ize again in the future

File: tests/integration_tests/build/test_binary_size.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    namespace = globals()
+    size_target_name = [name for name in namespace if namespace[name]
+                        is size_target][0]
+
+    # Compute concrete binary size difference.
+    delta_size = size_target - binary_size
+
+    binary_low_msg = (
+        'Current {} binary size of {} bytes is below the target of {} bytes with {} bytes.\n'
+        'Perhaps the value of {} should be updated'
```
Comment (aghecenco) [inline tests/integration_tests/build/test_binary_size.py]: nit
```suggestion
        'Update the {} threshold.'
```

File: tests/integration_tests/build/test_binary_size.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+import os
+import pytest
+import platform
+
+import host_tools.cargo_build as host
+
+MACHINE = platform.machine()
+""" Platform definition used to select the correct size target"""
+
+FC_BINARY_SIZE_TARGET = 3137280 if MACHINE == "x86_64" else 3536544
```
Comment (aghecenco) [inline tests/integration_tests/build/test_binary_size.py]: Interesting how it's significantly bigger on `aarch64`.

Comment (raduiliescu) [reply tests/integration_tests/build/test_binary_size.py]: Wouldn't a map per arch/features look better?

File: tests/integration_tests/build/test_binary_szie.py (hunk 6 above)
Comment (andrei-cosmin) [reply tests/integration_tests/build/test_binary_szie.py]: Done

File: tests/integration_tests/build/test_binary_szie.py (hunk 2 above)
Comment (andrei-cosmin) [reply tests/integration_tests/build/test_binary_szie.py]: Done

File: tests/integration_tests/build/test_binary_size.py (hunk 10 above)
Comment (andrei-cosmin) [reply tests/integration_tests/build/test_binary_size.py]: Done

File: tests/integration_tests/build/test_binary_szie.py (hunk 1 above)
Comment (andrei-cosmin) [reply tests/integration_tests/build/test_binary_szie.py]: Done

File: tests/integration_tests/build/test_binary_size.py (hunk 12 above)
Comment (dianpopa) [reply tests/integration_tests/build/test_binary_size.py]: > Interesting how it's significantly bigger on `aarch64`.

Off the top of my head, a bigger binary for aarch64 could have as its source the static linking to the libfdt library which enables the flatten device tree construction.

**Comment by dianpopa** [comment]
@aghecenco @raduiliescu @sandreim can you please check that your comments were addressed and resolve conversation where the case?

File: tests/integration_tests/build/test_binary_size.py (hunk 9 above)
Comment (andrei-cosmin) [reply tests/integration_tests/build/test_binary_size.py]: Done

**Pull request merged**